// @tres-erres/ngx-utils/esm5/lib/comprobante.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostListener } from '@angular/core';
var ComprobanteDirective = /** @class */ (function () {
    function ComprobanteDirective(el) {
        this.el = el;
        // Allow decimal numbers and negative values
        this.regex = new RegExp(/^[0-9]$/g);
        // Allow key codes for special events. Reflect :
        // Backspace, tab, end, home
        this.specialKeys = ['Backspace', 'Tab', 'End', 'Home', '-', 'ArrowLeft', 'ArrowRight'];
    }
    /**
     * @param {?} event
     * @return {?}
     */
    ComprobanteDirective.prototype.onKeyDown = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.specialKeys.indexOf(event.key) !== -1) {
            return;
        }
        /** @type {?} */
        var esNumero = String(event.key).match(this.regex);
        /** Si no es tecla especial o número */
        if (!esNumero && event.key !== 'Enter') {
            event.preventDefault();
            return;
        }
        /** @type {?} */
        var current = this.el.nativeElement.value;
        /** @type {?} */
        var next = this.el.nativeElement.value.concat(event.key);
        /**
         * Si está seleccionado, reemplazo
         * @type {?}
         */
        var seleccionado = this.el.nativeElement.selectionStart === 0 && this.el.nativeElement.selectionEnd === current.length && current.length > 0;
        if (seleccionado && esNumero) {
            event.target.value = event.key;
            event.preventDefault();
            return;
        }
        if (next.length > 13 && event.key !== 'Enter') {
            event.preventDefault();
            return;
        }
        if (event.key === 'Enter') {
            /** @type {?} */
            var guion = current.indexOf('-');
            if (guion === -1) {
                event.target.value = '0000-' + '0'.repeat(8 - current.length) + current;
            }
            else {
                /** @type {?} */
                var valor = current.replace('-', '');
                event.target.value = '0'.repeat(4 - valor.substring(0, guion).length) + valor.substring(0, guion) + '-' + '0'.repeat(8 - valor.substring(guion, valor.length).length) + valor.substring(guion, valor.length);
            }
        }
        else if (current.length === 4 && !current.includes('-')) {
            event.target.value = current + '-' + event.key;
            event.preventDefault();
        }
    };
    ComprobanteDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[trComprobante]'
                },] }
    ];
    /** @nocollapse */
    ComprobanteDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    ComprobanteDirective.propDecorators = {
        onKeyDown: [{ type: HostListener, args: ['keydown', ['$event'],] }]
    };
    return ComprobanteDirective;
}());
export { ComprobanteDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ComprobanteDirective.prototype.regex;
    /**
     * @type {?}
     * @private
     */
    ComprobanteDirective.prototype.specialKeys;
    /**
     * @type {?}
     * @private
     */
    ComprobanteDirective.prototype.el;
}
