// @tres-erres/ngx-utils/esm5/lib/importe.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostListener } from '@angular/core';
var ImporteDirective = /** @class */ (function () {
    function ImporteDirective(el) {
        this.el = el;
        // Allow decimal numbers and negative values
        this.regex = new RegExp(/^-?[0-9]+(\.[0-9]*){0,1}$/g);
        // Allow key codes for special events. Reflect :
        // Backspace, tab, end, home
        this.specialKeys = ['Backspace', 'Tab', 'End', 'Home', '-', 'ArrowLeft', 'ArrowRight'];
    }
    /**
     * @param {?} event
     * @return {?}
     */
    ImporteDirective.prototype.onKeyDown = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var current = this.el.nativeElement.value;
        // Autocompletar decimal
        if (event.key === '.') {
            if (current.length === 0) {
                event.target.value = '0.';
                event.preventDefault();
                return;
            }
            else if (current === '-') {
                event.target.value = '-0.';
                event.preventDefault();
                return;
            }
        }
        // No permitir otro menos que no sea al principio
        if (event.key === '-' && current.length > 0) {
            event.preventDefault();
        }
        // Allow Backspace, tab, end, and home keys
        if (this.specialKeys.indexOf(event.key) !== -1) {
            return;
        }
        // No permitir más de 2 decimales
        if (current.indexOf('.') !== -1) {
            /** @type {?} */
            var decimales = current.split('.')[1];
            if (decimales.length > 1) {
                event.preventDefault();
            }
        }
        /** @type {?} */
        var next = current.concat(event.key);
        if (next && !String(next).match(this.regex)) {
            event.preventDefault();
        }
    };
    ImporteDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[trImporte]'
                },] }
    ];
    /** @nocollapse */
    ImporteDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    ImporteDirective.propDecorators = {
        onKeyDown: [{ type: HostListener, args: ['keydown', ['$event'],] }]
    };
    return ImporteDirective;
}());
export { ImporteDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ImporteDirective.prototype.regex;
    /**
     * @type {?}
     * @private
     */
    ImporteDirective.prototype.specialKeys;
    /**
     * @type {?}
     * @private
     */
    ImporteDirective.prototype.el;
}
